### ISSUE TITLE
Shadow tree event processing model

### ISSUE THREAD

**Issue by annevk** [WICG/webcomponents#485]
I think if if you wanted to calculate this prior to dispatch you'd need a list of values for relatedTarget (and target), not just a single "relative related target" and "relative target".

**Comment by annevk** [comment]
I'm guessing that the simplest solution here is that for each node in the event path, we also store a target and a related target. Or is there a more clever way to approach this? @smaug----?

**Comment by annevk** [comment]
It also falls flat because the input is event's `currentTarget`, which is null prior to dispatch.

**Comment by annevk** [comment]
Ignoring `relatedTarget` and touch events for the moment, I have come up with this as a potential way to modify event dispatch taking into account multiple targets:

```
Event has an associated **path** (an ordered set).

Event has associated **targets** (an ordered set).

Dispatch(event, target, targetOverride):

  -> Append target to event's path.

  -> Append target to event's targets.

  -> Let parent be the result of invoking target's "get the parent" with event.

  -> While parent is non-null:

    -> Append parent to event's path.

    -> Append the result of "retargating" with parent and target to event's targets.

    -> Set parent to the result of parent's "get the parent" with event.
```

This does assume that https://w3c.github.io/webcomponents/spec/shadow/#dfn-retargeting-algorithm is correct, but it seems there should be a simpler way to calculate that without iterating over all target's ancestors.

**Comment by hayatoito** [comment]
I agree that the current spec is very lazy. I guess the usage of _target_  is ambiguous, right? It sometimes means the original _target_ value, and it sometime means retargeted _target_.

Yeah, we should clarify this by having another **targets**. That seems fine to me.

**Comment by annevk** [comment]
I don't really understand the current specification. It acknowledges you can have multiple relative targets and it says they have to be calculated prior to dispatch, but it never explains how that is done. And then in the dispatch algorithm itself it only talks about "relative target" as if it is a single thing.

It seems to me that if you can have multiple and if they need to be known prior to dispatch, we need to store them somewhere, just like we need to store the path.

Now, a question I had is if we write down that computation, whether it can be simpler than what I have above. Because for each "get the parent" invocation, we now invoke "retargating", which ends up iterating over target's ancestors. Is there no better way to do that?

**Comment by smaug----** [comment]
(not talking about relatedTarget here, just retargeting target)
Gecko ends up effectively iterate during capture phase, but it could be optimized out:
- When Gecko creates event target chain (== path), it goes from target up. Each item in the chain
  is stored as separate object, simplified as (currentTarget, newTarget)
  So normally it is enough to store currentTarget, but when dealing with node which parent in event path is in different subtree (so, crossing shadow boundary), also newTarget is stored for the _parent_ item. So during creation child tells whether parent needs newTarget.
- The first item in the chain has both currentTarget and newTarget stored, pointing to the same object. This just simplifies the implementation a bit.
- Now, when event is being handled, during _capture_ phase whenever we see that item has newTarget set, we need to retarget to something inside the anonymous tree. So the event target chain is iterated towards the original target and the first time newTarget is stored on some item, that is used as a new target. This is the case where we could possibly always store newTarget in item, and effectively retarget always, but retargeting would usually just end up changing .target to its current value. That would optimize out the iteration.
  (I can't recall why I did it the way I did 10 years ago :) )
- _target_ phase doesn't need any special handling
- _bubble_ phase is in Gecko simpler than capture, since there newTarget can be used without iteration. Whenever item has newTarget set to non-null, make event's target to be that one.

This all sounds very much like https://github.com/w3c/webcomponents/issues/485#issuecomment-211852259 : don't store just target, but some tuple (currenTarget, target) - and maybe also relatedTarget?.
https://github.com/w3c/webcomponents/issues/485#issuecomment-211914860 is perhaps a bit hard to follow. I'd rather have just one list containing these tuples.

**Comment by hayatoito** [comment]
Oh. Now I think I understand the issue. 

> The event target retargeting process must occur prior to dispatch of an event. In other words, any DOM mutation occurred in an event listener does not have any affect on the result of retargeting process.

We can ignore these sentences. These look obsolete. 

The current spec has started to use "get the parent", where we are actually doing _retargeting_.
Thus, the timing of _retargeting_ is now well defined. We no longer need these sentences.

I agree that we should _store_ a tuple of (relative target, relative related target) for each node, which is calculated at "get the parent", to clarify that `event.target` and `event.relatedTarget` return the value in the tuple.

BTW, the value of a tuple, (relative target, relative related target), is always the same in the same tree. However, I do not think the spec has to optimize it. Just storing a tuple in each Node is good enough, I think.

**Comment by annevk** [comment]
A list of tuples makes sense, yes. During the capture and bubble phases we should also switch the phase to target whenever we retarget, right?

**Comment by annevk** [comment]
Also, if an event does not bubble we still need to visit the different targets. So it might actually make sense to not always store the target, since you have to figure out when it's different anyway.

**Comment by annevk** [comment]
Okay, so how about this:

```
Event has an associated **path** (an ordered set of tuples).

Dispatch(event, target, targetOverride):

  -> Set event's dispatch flag.

  -> If targetOverride is not given, let targetOverride be target.

     Note: this is only used for a limited set of scenarios in HTML. We should 
     maybe simplify how it works since specifications must not abuse it.

  -> Append (target, targetOverride) to event's path.

  -> Let parent be the result of invoking target's "get the parent" with event.

  -> While parent is non-null:

    -> If target's root is a shadow-including inclusive ancestor of parent, 
       then append (parent, null) to event's path.

    -> Otherwise:

      -> Set target to parent.

      -> Append (parent, target) to event's path.

    -> Set parent to the result of parent's "get the parent" with event.

  -> For each _tuple_ in event's path, in reverse order:

    -> Set event's target to the target of the last tuple in event's path, 
       that is either _tuple_ or before _tuple_, whose target is non-null.

    -> If _tuple_'s target is non-null, then set event's phase to TARGET.

    -> Otherwise, set event's phase to CAPTURE.

    -> Invoke _tuple_'s item with event.

  -> If event's bubbles attribute is value is true: 

    -> Set event's phase to BUBBLE.

    -> For each _tuple_ in event's path, in order, if _tuple_'s target is null, 
       then invoke _tuple_'s item with event.

  -> Set event's phase to NONE.

  -> Set event's currentTarget to null.

  -> Return false if event's canceled flag is set, and true otherwise.
```

This does mingle the CAPTURE and TARGET phase, which I now realize Shadow DOM does differently (it seems to want to mingle TARGET and BUBBLE). Any reason Shadow DOM does what it does? Seems more natural to handle TARGET on the way down.

I also realized that the stop propagation flag should probably move to invoke since checking it each time we run invoke is rather redundant.

**Comment by annevk** [comment]
Also, another change I was thinking of making was to move some of the logic you put in "get the parent" in the dispatch algorithm instead. And make "get the parent" just responsible for returning the parent.

**Comment by smaug----** [comment]
> > If targetOverride is given, set event's target to targetOverride.

I think that is not quite right. load event uses targetOverride, right? and it should have AT_TARGET phase.

So bubble phase doesn't seem to handle retargeting. Or I guess the bubble phase handling is still missing the iteration of tuples or something.

**Comment by annevk** [comment]
Yeah, for the bubble phase we should set the target too. Either in the same way as capture does it, or the more optimized way of setting it when the target slot of the tuple is non-null.

**Comment by annevk** [comment]
Apart from that one other thing came up which @smaug---- and I discussed on IRC, when to invoke listeners:
1. capture target capture deeptarget bubble bubble (this is what I wrote up)
2. capture capture deeptarget bubble target bubble (this is what I think Shadow DOM defines)
3. capture target (only capture listeners) capture deeptarget bubble target (only bubble listeners) bubble

This last one is apparently something Gecko does today. It requires keeping additional state around (whether you're capture or bubble) and could make shadow trees observable since if you add event listeners for capture, then bubble, then capture, you can tell that when the two for capture have been invoked the one for bubble hasn't yet been.

**Comment by annevk** [comment]
@rniwa opinions on any of this?

**Comment by annevk** [comment]
It seems @smaug---- separated that out into https://github.com/whatwg/dom/issues/237 so maybe feedback should go there instead.
